google/cloud/spanner/keys.h

```diff
@@ -202,8 +246,65 @@ class KeySet {
   bool IsAll() const { return all_; }
 
  private:
+  template <typename RowType>
+  friend class KeySetBuilder;
+
+  friend ::google::spanner::v1::KeySet internal::ToProto(KeySet const& keyset);
+
   explicit KeySet(bool all) : all_(all) {}
 
+  class ValueRow {
+   public:
+    template <std::size_t SIZE>
+    explicit ValueRow(std::array<Value, SIZE> const& column_values)
+        : column_values_(
+              std::vector<Value>(column_values.begin(), column_values.end())) {}
+
+    std::vector<Value> const& column_values() const { return column_values_; }
+
+   private:
+    std::vector<Value> column_values_;
+  };
+
+  class ValueBound {
+   public:
+    enum class Mode { MODE_CLOSED, MODE_OPEN };
+    explicit ValueBound(ValueRow key, ValueBound::Mode mode)
+        : key_(std::move(key)), mode_(mode) {}
+
+    ValueRow const& key() const { return key_; }
+    Mode const& mode() const { return mode_; }
+
+   private:
+    ValueRow key_;
+    Mode mode_;
+  };
+
+  class ValueKeyRange {
+   public:
+    explicit ValueKeyRange(ValueBound start, ValueBound end)
+        : start_(std::move(start)), end_(std::move(end)) {}
+
+    ValueBound const& start() const { return start_; }
+    ValueBound const& end() const { return end_; }
+
+   private:
+    ValueBound start_;
+    ValueBound end_;
+  };
+
+  KeySet& EmplaceKey(ValueRow key) {
```

Comment:
I think all of the "emplace" things in this file should really be "push_back", or in the vocabulary of the KeySet API, maybe name them `AddRow`.

The reason is that "emplace" methods are mostly useful when constructing an object in place from its ingredients. Like, `persons.emplace_back("Jerry", "Seinfeld", Occupation::kComedian)`. However, in all of the cases in this file, we have an already constructed object and we're simply adding it to the container. `vector::push_back` is already aware of rvalues and is move-efficient, so I don't think there's any performance gain to be had with emplacing.

See also https://abseil.io/tips/112

Reply:
Done.